ssl/ssl_cert.cc

```diff
@@ -300,49 +319,55 @@ static int cert_set_chain_and_key(
     }
   }
 
-  cert->privatekey = UpRef(privkey);
+  int idx = cert->cert_privatekey_idx;
+  cert->cert_privatekeys[idx].privatekey = UpRef(privkey);
   cert->key_method = privkey_method;
 
-  cert->chain = std::move(certs_sk);
+  cert->cert_privatekeys[idx].chain = std::move(certs_sk);
   return 1;
 }
 
 bool ssl_set_cert(CERT *cert, UniquePtr<CRYPTO_BUFFER> buffer) {
-  switch (check_leaf_cert_and_privkey(buffer.get(), cert->privatekey.get())) {
+  CERT_PKEY &cert_pkey = cert->cert_privatekeys[cert->cert_privatekey_idx];
+
+  switch (
+      check_leaf_cert_and_privkey(buffer.get(), cert_pkey.privatekey.get())) {
     case leaf_cert_and_privkey_error:
       return false;
     case leaf_cert_and_privkey_mismatch:
       // don't fail for a cert/key mismatch, just free current private key
       // (when switching to a different cert & key, first this function should
       // be used, then |ssl_set_pkey|.
-      cert->privatekey.reset();
+      cert_pkey.privatekey.reset();
       break;
     case leaf_cert_and_privkey_ok:
       break;
   }
 
-  if (cert->chain != nullptr) {
-    CRYPTO_BUFFER_free(sk_CRYPTO_BUFFER_value(cert->chain.get(), 0));
-    sk_CRYPTO_BUFFER_set(cert->chain.get(), 0, buffer.release());
+  if (cert_pkey.chain != nullptr) {
+    CRYPTO_BUFFER_free(sk_CRYPTO_BUFFER_value(cert_pkey.chain.get(), 0));
+    sk_CRYPTO_BUFFER_set(cert_pkey.chain.get(), 0, buffer.release());
     return true;
   }
 
-  cert->chain.reset(sk_CRYPTO_BUFFER_new_null());
-  if (cert->chain == nullptr) {
+  cert_pkey.chain.reset(sk_CRYPTO_BUFFER_new_null());
+  if (cert_pkey.chain == nullptr) {
     return false;
   }
 
-  if (!PushToStack(cert->chain.get(), std::move(buffer))) {
-    cert->chain.reset();
+  if (!PushToStack(cert_pkey.chain.get(), std::move(buffer))) {
+    cert_pkey.chain.reset();
     return false;
   }
 
   return true;
 }
 
 bool ssl_has_certificate(const SSL_HANDSHAKE *hs) {
-  return hs->config->cert->chain != nullptr &&
-         sk_CRYPTO_BUFFER_value(hs->config->cert->chain.get(), 0) != nullptr &&
+  CERT_PKEY &cert_pkey =
+      hs->config->cert->cert_privatekeys[hs->config->cert->cert_privatekey_idx];
+  return cert_pkey.chain != nullptr &&
+         sk_CRYPTO_BUFFER_value(cert_pkey.chain.get(), 0) != nullptr &&
          ssl_has_private_key(hs);
 }
```

Comment:
My impression is that almost everywhere that `cert_privatekeys` is accessed by an index needs to be guarded by a  `if(cert->cert_privatekeys.size() < 1) { /* error handling and return */ }`
```
bool ssl_has_certificate(const SSL_HANDSHAKE *hs) {
  CERT* cert = hs->config->cert.get();
  if(cert->cert_privatekeys.size() < 1) {
      return false;
  }
  CERT_PKEY &cert_pkey =
      cert->cert_privatekeys[cert->cert_privatekey_idx];
  return cert_pkey.chain != nullptr &&
         sk_CRYPTO_BUFFER_value(cert_pkey.chain.get(), 0) != nullptr &&
         ssl_has_private_key(hs);
}
```